modules/core/src/main/java/org/apache/ignite/internal/processors/metastorage/persistence/DistributedMetaStorageImpl.java

```diff
@@ -1361,14 +1396,75 @@ private RuntimeException criticalError(Throwable e) {
         throw U.convertException((IgniteCheckedException)e);
     }
 
+    /**
+     * Store data in local metastorage or in memory and log with given prefix. Doesn't optimise history items.
+     *
+     * @param prefix Prefix for log message, should end with '['.
+     * @param items Items to store.
+     * @param index Starting index.
+     */
+    private void completeWriteAndLog(String prefix, DistributedMetaStorageHistoryItem[] items, int index) {
+        StringBuilder sb = new StringBuilder(prefix);
+
+        sb.append("updates=[");
+
+        DistributedMetaStorageVersion initialVer = ver;
+
+        for (int i = index; i < items.length; i++) {
+            try {
+                completeWrite(items[index], false);
+
+                sb.append(printHistoryItem(items[index]));
+            } catch (IgniteCheckedException ex) {
+                log.error("Unable to unmarshal new metastore data. update=" + items[index], ex);
+            }
+
+            if (i < items.length - 1)
+                sb.append(", ");
+        }
+
+        sb.append("], initialVer=")
+            .append(initialVer)
+            .append(", ver=")
+            .append(ver)
+            .append("]");
+
+        log.info(sb.toString());
+    }
+
+
+    /**
+     * Store data in local metastorage or in memory and log. Removes history item values that match already existing
+     * values. Might lead to no-op.
+     *
+     * @param histItem {@code <key, value>} pair to process.
+     * @throws IgniteCheckedException In case of IO/unmarshalling errors.
+     */
+    private void completeWriteAndLog(DistributedMetaStorageHistoryItem histItem) throws IgniteCheckedException {
```

Comment:
```suggestion
    private void completeWriteAndLog(DistributedMetaStorageHistoryItem histItem) throws IgniteCheckedException {
```
There are 2 methods which are called very similarly,  but they pass different values of `optimize`. This is confusing. Should we adjust naming to make this difference obvious?